src/software/amazon/ion/impl/LocalSymbolTableImportAdapter.java

```diff
@@ -0,0 +1,243 @@
+            readOnlyDelegate.makeReadOnly();
+        }
+
+        return new LocalSymbolTableImportAdapter(readOnlyDelegate);
+    }
+
+    private LocalSymbolTableImportAdapter(final LocalSymbolTable delegate)
+    {
+        super(delegate);
+
+        imports = createImportList(delegate);
+
+        systemSymbolCount = countSystemSymbols(getDelegate());
+        maxImportedId = getDelegate().getImportedMaxId() - systemSymbolCount;
+        maxId = getDelegate().getMaxId() - systemSymbolCount;
+        numberOfLocalSymbols = maxId - maxImportedId;
+    }
+
+    @Override
+    public boolean isReadOnly()
+    {
+        return true;
+    }
+
+    @Override
+    public void makeReadOnly()
+    {
+        // read only by default
+    }
+
+    @Override
+    public SymbolTable getSystemSymbolTable()
+    {
+        return null;
+    }
+
+    @Override
+    public SymbolTable[] getImportedTables()
+    {
+        if (importedTables == null)
+        {
+            importedTables = imports.toArray(new SymbolTable[imports.size()]);
+        }
+
+        return importedTables;
+    }
+
+    @Override
+    public int getImportedMaxId()
+    {
+        return maxImportedId;
+    }
+
+    @Override
+    public int getMaxId()
+    {
+        return maxId;
+    }
+
+    @Override
+    public SymbolToken intern(String text)
+    {
+        throw new ReadOnlyValueException(this.getClass());
+    }
+
+    @Override
+    public SymbolToken find(String text)
+    {
+        for (SymbolTable importedTable : imports)
+        {
+            SymbolToken symbolToken = importedTable.find(text);
+
+            if (symbolToken != null)
+            {
+                return symbolToken;
+            }
+        }
+
+        Integer sid = getLocalTextToSidMap().get(text);
+        if (sid != null)
+        {
+            return new SymbolTokenImpl(text, sid);
+        }
+
+        return null;
+    }
+
+    @Override
+    public int findSymbol(String name)
+    {
+        SymbolToken token = find(name);
+        return (token == null ? UNKNOWN_SYMBOL_ID : token.getSid());
+    }
+
+    @Override
+    public String findKnownSymbol(int id)
+    {
+        if (id < 0)
+        {
+            throw new IllegalArgumentException("symbol IDs must be greater than 0");
+        }
+
+        if (id > getMaxId()) // not in this table or imports
+        {
+            return null;
+        }
+
+        // look into imports
+        if (id <= maxImportedId)
+        {
+            for (SymbolTable importedSymbolTable : imports)
+            {
+                String text = importedSymbolTable.findKnownSymbol(id);
+                if (text != null)
+                {
+                    return text;
+                }
+            }
+        }
+
+        // not in imports get from local
+        int index = id - maxImportedId - 1;
+        return getSidToLocalText()[index];
+    }
+
+    private List<SymbolTable> createImportList(LocalSymbolTable delegate)
```

Comment:
Why does the adapter need to create its own imports list and symbols maps? Can't it just call through to the delegate's `find*()` methods?

Reply:
It can't, the main reason for the adapter is to remove the `SystemSymbolTable (SST)` from the delegate import list to avoid having it twice, one in the imported and on in the importing LST. 

SIDs are scoped locally when delegating to imports, e.g. 

* LST_A = `{a1, a2}`,  `a1.sid = 10` and `a2.sid=11`
* LST_B = `{b1}`  and imports LST_A so `b.sid = 12`

`LST_B.findKnownSymbol(10)` triggers a call to `LST_A.findKnownSymbol(1)`, without the adapter, or passing through to the delegate, this would return `$ion` instead of `a1`

This scoping is done in `LocalSymbolTableImports.findKnownSymbol`

Reply:
Got it. My motivation here is that the logic looks similar to existing lookup logic; it would be nice to be able to reuse existing logic if possible. I'm just brainstorming here.

It looks like the system symbol table is guaranteed to be every local symbol table's first import, correct? In that case, a `LocalSymbolTableImportAdapter`'s system symbol table should never be consulted; the owning table's system symbol table should always be consulted first. Is there a way we can use that knowledge to simplify the `LocalSymbolTableImportAdapter`? Can the adapter simply add the max system SID to each requested SID so that the SID falls within the delegate's local symbols range, eliminating the need to keep separate imports and symbols lists in the adapter?

From your example, the user calls `LST_B.findKnownSymbol(10)`, which `LocalSymbolTableImports.findKnownSymbol` determines falls within LST_B's second import (LST_A, right behind the system symbol table). That's when `LocalSymbolTableImportAdapter(LST_A).findKnownSymbol(1)` is called, right? Can the adapter simply add `countSystemSymbols()` to the input SID so that it falls within LST_A's local symbols range, then delegate directly to `LST_A.findKnownSymbol`, rather than keeping a copy of LST_A's imports within the adapter?

Reply:
Hum good idea, will try it out